File: Engines/scxml4cpp/engine/src/Event.cpp
/*
 *    scampl4cpp/engine
 *
 *    Copyright by European Southern Observatory, 2012
 *    All rights reserved
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 2.1 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with this library; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 
 *    02111-1307 USA.
 */

/*
 * $Id: Event.cpp 1061 2015-07-13 15:03:59Z landolfa $
 */

#include "Event.h"
#include "Log.h"

using namespace scxml4cpp;

Event::Event() :
    mId(""),
    mType(Event::CHANGE_EVENT),
    mStatus(Event::TOBEPROCESSED),
    mPayload(NULL)
{
    FILE_LOG(logTRACE);
}

Event::Event(const std::string& id, EventType type) :
    mId(id),
    mType(type),
    mStatus(Event::TOBEPROCESSED),
    mPayload(NULL)
{    
    FILE_LOG(logTRACE);
}

/*
 * Copy constructor
 */
Event::Event(const Event& e) :
    mId(e.getId()),
    mType(e.getType()),
    mStatus(e.getStatus()),
    mPayload(e.getPayload())
{    
    FILE_LOG(logTRACE);
}

/*
 * Assignment operator
 */
Event& Event::operator=(const Event& e) 
{    
    FILE_LOG(logTRACE);
    mId = e.getId();
    mType = e.getType();
    mStatus = e.getStatus();
    mPayload = e.getPayload();
    return *this;
}


Event::~Event()
{    
    FILE_LOG(logTRACE);
}

std::string Event::getId() const
{    
    FILE_LOG(logTRACE);
    return mId;
}

Event::EventType Event::getType() const
{    
    FILE_LOG(logTRACE);
    return mType;
}

Event::EventStatus Event::getStatus() const
{    
    FILE_LOG(logTRACE);
    return mStatus;
}

void* Event::getPayload() const
{    
    FILE_LOG(logTRACE);
    return mPayload;
}

void Event::setId(std::string id)
{    
    FILE_LOG(logTRACE);
    mId = id;
}

void Event::setType(EventType type)
{    
    FILE_LOG(logTRACE);
    mType = type;
}

void Event::setStatus(EventStatus status)
{    
    FILE_LOG(logTRACE);
    mStatus = status;
}

void Event::setPayload(void* payload)
{    
    FILE_LOG(logTRACE);
    mPayload = payload;
}

bool Event::isNull() const
{
    FILE_LOG(logTRACE);
    return mId.empty();
}

File: Engines/scxml4py/src/scxml4py/event.py
'''
    event module part of scxml4py.
    
    @authors: landolfa
    @date: 2016-12-26
    
    @copyright: LGPL 2.1 
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 
    02111-1307 USA.
'''

'''
    $Id: event.py 1061 2015-07-13 15:03:59Z landolfa $
'''

from enum import Enum
from functools import total_ordering


class EventType(Enum):
    CALL_EVENT = 0 
    CHANGE_EVENT = 1
    SIGNAL_EVENT = 2
    TIME_EVENT = 3
    ERROR_EVENT = 4

class EventStatus(Enum):
    TOBEPROCESSED = 0 
    PROCESSED = 1
    IGNORED = 2
    REJECTED = 3
    DEFERRED = 4

@total_ordering
class Event(object):

    def __init__(self, theId, theType = EventType.CHANGE_EVENT, theStatus = EventStatus.TOBEPROCESSED):
        self.mId = theId
        self.mType = theType
        self.mStatus = theStatus
        
    def __str__(self):
        return self.mId.__str__()

    def __lt__(self, other):
        if other == None:
            return False        
        return self.mId.__str__() < other.mId.__str__()

    def __eq__(self, other):
        if other == None:
            return False        
        return self.mId.__str__() == other.mId.__str__()

    def getId(self):
        return self.mId
        
    def getType(self, theType):
        self.mType = theType
        
    def getStatus(self, theStatus):
        self.mStatus = theStatus
            
    def setId(self, theId):
        self.mId = theId
    
    def setType(self, theType):
        assert (theType >= EventType.CALL_EVENT and theType <= EventType.ERROR_EVENT)
        self.mType = theType

    def setStatus(self, theStatus):
        # @TODO assertion doesn't work:
        # "TypeError: unorderable types: EventStatus() >= EventStatus()"
        #assert (theStatus >= EventStatus.TOBEPROCESSED and theStatus <= EventStatus.DEFERRED)
        self.mStatus = theStatus

    def clone(self, regionNo):
        clonedEvent = Event(self.mId + str(regionNo))
        clonedEvent.mType = self.mType
        # runtime info does not need to be cloned
        #clonedEvent.mStatus = self.mStatus
        return clonedEvent
    